### PR TITLE
feat: introduce table component for next theme

### PR DESCRIPTION
**Pull request by dcshzj**
Figma: https://www.figma.com/file/8TEzSSDbx97gctRJ7nKdqG/Website-Template-Gen-2?type=design&node-id=2445-6145&mode=design&t=sKnfu9ZnyCEdcQ5R-4

Not done yet, will do separately:
- Empty file prepended for links to documents
- Right-alignment of numbers and numbers-like strings (i.e. currency)

### CODE REVIEW

**Comment by jacobkwan** [comment]
!run chromatic

File: src/templates/next/components/Table/Table.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import BaseParagraph from "../shared/Paragraph"
+import Image from "../Image"
+import OrderedList from "../OrderedList"
+import UnorderedList from "../UnorderedList"
+import { Caption } from "../../typography/Caption"
+
+const getIsTableOverflowing = (tableRef: React.RefObject<HTMLTableElement>) => {
+  const tableRight = tableRef.current?.getBoundingClientRect().right || 0
+  return tableRight > window.innerWidth
+}
```
[the remark this answers is not in the document]
Comment (dcshzj) [reply src/templates/next/components/Table/Table.tsx]: Right, you make a really good point here. Thought it worked cos I was only testing within storybook for this single component.

Updated in fda45db71bdcda84a532c5465f1c4959d484c3cf.

File: src/templates/next/components/Table/Table.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+const Table = ({ items, caption, LinkComponent = "a" }: TableProps) => {
```
[the remark this answers is not in the document]
Comment (dcshzj) [reply src/templates/next/components/Table/Table.tsx]: Removed in f44ca0d50c2376b86eb986bd1eabca8906b93242.

File: src/common/Table.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  value: (string | ImageProps | OrderedListProps | UnorderedListProps)[]
+  colSpan?: number
+  rowSpan?: number
+}
+
+interface TableCell extends TableBaseCell {
+  type: "tableCell"
+}
+
+interface TableHeaderCell extends TableBaseCell {
```
[the remark this answers is not in the document]
Comment (jacobkwan) [reply src/common/Table.ts:15]: that table is hideous LOL but okay understand its needed

File: src/templates/next/components/Table/Table.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (jacobkwan) [reply src/templates/next/components/Table/Table.tsx]: can help to remove from `TableProps` too, tq

File: src/templates/next/components/Table/Table.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  const [isTableOverflowing, setIsTableOverflowing] = useState(false)
+  const tableRef = useRef<HTMLTableElement>(null)
+  // Generate a random ID for the caption that is sufficiently unique
+  const captionId = `caption-${Math.random()}`
```
[the remark this answers is not in the document]
Comment (jacobkwan) [reply src/templates/next/components/Table/Table.tsx:21]: yup understand it's there for accessibility but just seems weird that the `id` is not a user-friendly string derived from the table caption or smth

like if there are multiple tables in the page, then you have `<p id=caption-0.123123>`, `<p id=caption-0.5394978`, `<p id=caption-0.82831` and the corresponding `aria-describedby`s 🤣 

no strong opinion here so can leave it also

File: src/templates/next/layouts/Content/Content.stories.tsx
Comment (jacobkwan) [inline src/templates/next/layouts/Content/Content.stories.tsx:1]: rip it looks really good but there's some issues around ~1000px screen width
the max-w and padding from the Content layout seems to have some weird interaction with Table
I think it's due to the `min-w-40` on each cell

not sure why its not keeping to the container and becoming scrollable but instead eating into the gutters

File: src/templates/next/components/Table/Table.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (dcshzj) [reply src/templates/next/components/Table/Table.tsx]: Removed in 6d1734e6a12b737920fb71551a8bafdaae8a9fbe.

File: src/templates/next/components/Table/Table.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (dcshzj) [reply src/templates/next/components/Table/Table.tsx:21]: Yeah that was the intention, it's not intended to be user-friendly cos it's not used by the user directly. We can't guarantee that the caption content is unique within the page.

File: src/templates/next/layouts/Content/Content.stories.tsx
Comment (jacobkwan) [reply src/templates/next/layouts/Content/Content.stories.tsx:1]: also might have to add the `[&:not(:first-child)]:mt-xxx` to `Table`, looks jank if its the first thing right after a Header

Comment (dcshzj) [reply src/templates/next/layouts/Content/Content.stories.tsx:1]: > also might have to add the [&:not(:first-child)]:mt-xxx to Table, looks jank if its the first thing right after a Header

Do you have a screenshot for this?

Comment (jacobkwan) [reply src/templates/next/layouts/Content/Content.stories.tsx:1]: Screenshot 2024-04-09 at 3 06 13 PM

Comment (jacobkwan) [reply src/templates/next/layouts/Content/Content.stories.tsx:1]: > also might have some sane margin top for `Table` using the `[&:not(:first-child)]:mt-xxx`, looks jank if its the first thing right after a Header

just clarifying that this is what we are doing at the moment for the other Content page blocks like Headings

File: src/templates/next/components/Table/Table.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (dcshzj) [reply src/templates/next/components/Table/Table.tsx]: Turns out there's such a thing as `<caption>` for tables, switched to that in 55fe27d4efeb0775c8fbc52c08a921dd6770071d.

File: src/templates/next/layouts/Content/Content.stories.tsx
Comment (dcshzj) [reply src/templates/next/layouts/Content/Content.stories.tsx:1]: Adjusted in 342f365ac3963e2bf45ceb4c89a575a0ff20cbc0.

**Review by jacobkwan — APPROVED** [review]

**Pull request merged**